exe/Main.hs

```diff
@@ -188,22 +192,256 @@ showEvent lock (EventFileDiagnostics (toNormalizedFilePath' -> file) diags) =
+              -- A special target for the file which caused this wonderful
+              -- component to be created.
+              let special_target = (cfp, res)
+              --pprTraceM "TARGETS" (ppr (map (text . show) ctargets))
+              let xs = map (,res) ctargets
+              return (special_target:xs, res)
+
+        -- New HscEnv for the component in question
+        (cs, res) <- new_cache new
+        -- Modified cache targets for everything else in the hie.yaml file
+        -- which now uses the same EPS and so on
+        cached_targets <- concatMapM (fmap fst . new_cache) old_deps
+        modifyVar_ fileToFlags $ \var -> do
+            pure $ Map.insert hieYaml (HM.fromList (cs ++ cached_targets)) var
+
+        return res
+
+    lock <- newLock
+
+    -- This caches the mapping from hie.yaml + Mod.hs -> [String]
+    sessionOpts <- return $ \(hieYaml, file) -> do
+
+
+        fm <- readVar fileToFlags
+        let mv = Map.lookup hieYaml fm
+        let v = fromMaybe HM.empty mv
+        cfp <- liftIO $ canonicalizePath file
+        case HM.lookup (toNormalizedFilePath' cfp) v of
+          Just (_, old_di) -> do
+            deps_ok <- checkDependencyInfo old_di
+            unless deps_ok $ do
+              modifyVar_ fileToFlags (const (return Map.empty))
+              -- Keep the same name cache
+              modifyVar_ hscEnvs (return . Map.adjust (\(h, _) -> (h, [])) hieYaml )
+          Nothing -> return ()
+        -- We sort so exact matches come first.
+        case HM.lookup (toNormalizedFilePath' cfp) v of
+            Just opts -> do
+                --putStrLn $ "Cached component of " <> show file
+                pure (fst opts)
+            Nothing-> do
+                finished_barrier <- newBarrier
+                -- fork a new thread here which won't be killed by shake
+                -- throwing an async exception
+                void $ forkIO $ do
+                  putStrLn $ "Consulting the cradle for " <> show file
+                  cradle <- maybe (loadImplicitCradle $ addTrailingPathSeparator dir) loadCradle hieYaml
+                  opts <- cradleToSessionOpts cradle cfp
+                  print opts
+                  res <- fst <$> session (hieYaml, toNormalizedFilePath' cfp, opts)
+                  signalBarrier finished_barrier res
+                waitBarrier finished_barrier
+
+    dummyAs <- async $ return (error "Uninitialised")
+    runningCradle <- newIORef dummyAs
+    -- The main function which gets options for a file. We only want one of these running
+    -- at a time.
+    let getOptions file = do
+            hieYaml <- cradleLoc file
+            sessionOpts (hieYaml, file)
+    -- The lock is on the `runningCradle` resource
+    return $ \file -> liftIO $ withLock lock $ do
+        as <- readIORef runningCradle
+        finished <- poll as
+        case finished of
+            Just {} -> do
+                as <- async $ getOptions file
```

Comment:
It looks like you can get interrupted between creating the async and updating the IORef. Should this be masked?

Reply:
Yes good point.